org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs

```diff
@@ -25,49 +28,67 @@ public class NearInteractionModeDetector : ProximityDetector
         /// colliders stopped being detected and update their buttons front plate and rounded
         /// rect if they are labeled as dynamic (based on proximity).
         /// </summary>
-        List<Collider> previouslyDetectedColliders = new List<Collider>();
+        private List<Collider> previouslyDetectedColliders = new List<Collider>();
 
         /// <inheritdoc />
         public override bool IsModeDetected()
         {
             bool result = base.IsModeDetected() || IsNearInteractorSelecting();
             if (result)
             {
-                for (int i = 0; i < previouslyDetectedColliders.Count; i++)
+                UpdateProximityExited();
+                UpdateProximityEntered();
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Call OnProximityExited for all colliders that were previously detected but are no longer detected.  This
+        /// effectively triggers OnProximityExited for all colliders that are no longer detected.
+        /// </summary>
+        private void UpdateProximityExited()
+        {
+            for (int i = 0; i < previouslyDetectedColliders.Count; i++)
+            {
+                Collider previouslyDetectedCollider = previouslyDetectedColliders[i];
+                if (!DetectedColliders.Contains(previouslyDetectedCollider) &&
+                    InteractionManager.TryGetInteractableForCollider(previouslyDetectedCollider, out IXRInteractable xrInteractable) &&
+                    xrInteractable is IXRProximityInteractable xrProximityInteractable)
                 {
-                    Collider previouslyDetectedCollider = previouslyDetectedColliders[i];
-                    if (!DetectedColliders.Contains(previouslyDetectedCollider) && previouslyDetectedCollider != null)
+                    foreach (XRBaseInteractor xrBaseInteractor in nearInteractors)
                     {
-                        IXRProximityInteractable nearInteractionMode = previouslyDetectedCollider.GetComponent<IXRProximityInteractable>();
-                        if (nearInteractionMode != null)
-                        {
-                            foreach (XRBaseInteractor xrBaseInteractor in nearInteractors)
-                            {
-                                previouslyDetectedCollider.GetComponent<IXRProximityInteractable>().OnProximityExited(new ProximityExitedEventArgs(this, previouslyDetectedCollider, xrBaseInteractor));
-                            }
-                        }
-                        previouslyDetectedColliders.Remove(previouslyDetectedCollider);
+                        xrProximityInteractable.OnProximityExited(new ProximityExitedEventArgs(previouslyDetectedCollider, xrBaseInteractor));
                     }
+                    previouslyDetectedColliders.Remove(previouslyDetectedCollider);
```

Comment:
I still really, really don't think we should be removing from `previouslyDetectedColliders` while we're actively iterating through it. I don't see https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/611#discussion_r1489859824 addressed here, but maybe I'm misreading the change. (I think my previous comment may have been posted on the wrong loop)

Reply:
@ms-RistoRK ....If the foreach loop was changed to a for loop, and it iterates backwards, removal should avoid "skipping" colliders.

Reply:
Actually, that's exactly why I had two collections at the very first version of this PR.  One collection would be used track which (Colllider, Interactor) stopped ProximityHovering on the interactable, the second collection was used to track which (Collider, Interactor) start ProximityHovering on the interactable.  After those collections were update in a frame, then in the next frame, the collections would be first checked for changes and then trigger the corresponding OnProximityHover{Enter|Exit} events.  I was precisely trying to prevent the traversal of a collection as it could still be changed.  Then I lost track of that approach as I implemented other feedback.

Reply:
I think you are still missing colliders because you are not traversing backwards on `previouslyDetectedColliders`

`for (int i = 0; i < previouslyDetectedColliders.Count; i++)`

should be

`for (int i = previouslyDetectedColliders.Count; i >= 0; i--)`

Reply:
Implemented in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/659/commits/b4533fb2ae5a07724adf0bad32700d1ef82ea469